### PR TITLE
Interpretation templates

### PR DESCRIPTION
**Pull request by xispa**
## Description of the issue/feature this PR addresses

This Pull Requests adds the "Interpretation Templates" content type. User can add there free text, that can be used later to feed the results interpretation field from Sample. Above the Results Interpretation field in Sample view, a selector has been added for the user to be able to choose the template.

Captura de 2020-09-23 16-53-09

Captura de 2020-09-23 16-59-47

Captura de 2020-09-23 16-59-59

## Current behavior before PR

User has to always re-type or copy&paste pre-defined texts in the results interpretation field

## Desired behavior after PR is merged

User can maintain lab predefined texts in SENAITE and insert them as desired in the results interpretation field

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

### CODE REVIEW

File: src/bika/lims/browser/analysisrequest/templates/analysisrequest_results_interpretation.pt [hunk 1]
```diff
@@ -51,7 +51,36 @@
...
+                <tal:itemplates repeat="template view/get_interpretation_templates">
+                  <option tal:attributes="value template/uid"
+                          tal:content="template/title"/>
+                </tal:itemplates>
+              </select>
+
+              <button id="interpretationtemplate-insert"
+                      name="interpretationtemplate-insert"
+                      type="submit"
+                      class="btn allowMultiSubmit btn-outline-primary btn-sm"
```
Comment (ramonski) [inline src/bika/lims/browser/analysisrequest/templates/analysisrequest_results_interpretation.pt]: I think this `allowMultiSubmit` CSS class can be dropped or?

File: src/senaite/core/browser/content/interpretationtemplates.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        """Update hook
+        """
+        super(InterpretationTemplatesView, self).update()
+
+    def before_render(self):
+        """Before template render hook
+        """
+        super(InterpretationTemplatesView, self).before_render()
+        # Don't allow any context actions
+        self.request.set("disable_border", 1)
```
Comment (ramonski) [inline src/senaite/core/browser/content/interpretationtemplates.py]: You can also make the content implement `IHideActionsMenu` (`from senaite.core.interfaces import IHideActionsMenu`)

File: src/senaite/core/profiles/default/types/InterpretationTemplate.xml [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+<object name="DynamicAnalysisSpec" meta_type="Dexterity FTI"
+        i18n:domain="senaite.core"
+        xmlns:i18n="http://xml.zope.org/namespaces/i18n">
+
+  <!-- Title and Description -->
+  <property name="title" i18n:translate="">Interpretation Template</property>
+  <property name="description" i18n:translate=""/>
+
+  <!-- content-type icon -->
+  <property name="icon_expr">string:${portal_url}/++plone++senaite.core.static/images/interpretation_template.svg</property>
```
Comment (ramonski) [inline src/senaite/core/profiles/default/types/InterpretationTemplate.xml]: better use `senaite_theme/icon/interpretation_template`

File: src/senaite/core/profiles/default/types/InterpretationTemplates.xml [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+<object name="DynamicAnalysisSpecs" meta_type="Dexterity FTI"
+        i18n:domain="senaite.core"
+        xmlns:i18n="http://xml.zope.org/namespaces/i18n">
+
+  <!-- Title and Description -->
+  <property name="title" i18n:translate="">Interpretation Templates</property>
+  <property name="description" i18n:translate=""/>
+
+  <!-- content-type icon -->
+  <property name="icon_expr">string:${portal_url}/++plone++senaite.core.static/images/interpretation_template.svg</property>
```
Comment (ramonski) [inline src/senaite/core/profiles/default/types/InterpretationTemplates.xml]: better use `senaite_theme/icon/interpretation_template`

File: src/senaite/core/content/interpretationtemplates.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+from zope import schema
+from zope.interface import implementer
+
+
+class IInterpretationTemplates(model.Schema):
+    """Results Interpretation Templates folder interface
+    """
+    title = schema.TextLine(
+        title=_PMF(u'label_title', default=u'Title'),
+        required=True)
```
Comment (ramonski) [inline src/senaite/core/content/interpretationtemplates.py]: Shall we better  use the `plone.app.dexterity.behaviors.metadata.IBasic` behavior for those types?

**Review by ramonski — CHANGES_REQUESTED** [review]

**Comment by ramonski** [comment]
Tests are also failing 🤔

File: src/senaite/core/content/interpretationtemplates.py (hunk 5 above)
Comment (xispa) [reply src/senaite/core/content/interpretationtemplates.py]: I thought about this, but the `IBasic` behavior comes with other fields like description, etc.. and I thought that was better to stick to the minimum requirements. We want small objects, less overhead. What we can do is to create our own behavior and apply this behavior to all content "folders".

File: src/senaite/core/browser/content/interpretationtemplates.py (hunk 2 above)
Comment (xispa) [reply src/senaite/core/browser/content/interpretationtemplates.py]: Good point! thanks

File: src/bika/lims/browser/analysisrequest/templates/analysisrequest_results_interpretation.pt (hunk 1 above)
Comment (xispa) [reply src/bika/lims/browser/analysisrequest/templates/analysisrequest_results_interpretation.pt]: Yes, I think so

**Comment by xispa** [comment]
> Tests are also failing thinking

Maybe because I haven't added "InterpretationTemplate" into this rubbish?: https://github.com/senaite/senaite.core/blob/2.x/src/bika/lims/__init__.py#L68-L168 However, I can install a new site and it works w/o problems

**Comment by ramonski** [comment]
> > Tests are also failing thinking
> 
> Maybe because I haven't added "InterpretationTemplate" into this rubbish?: https://github.com/senaite/senaite.core/blob/2.x/src/bika/lims/__init__.py#L68-L168 However, I can install a new site and it works w/o problems

No, I don't think so. At least this is not what was needed for the dynamic analysis specs.
I think we need to add an dependency step in our GS profile like this:

```xml
  <!-- Install Handler -->
  <genericsetup:importStep
      name="senaite.core.setuphandlers.install"
      title="SENAITE CORE: Run Setup Handler"
      description="Run install handler"
      handler="senaite.core.setuphandlers.install">
    <depends name="typeinfo"/>
    <depends name="toolset"/>
  </genericsetup:importStep>
```


Because the setuphandler runs this one from the `bika` package explicitly:
https://github.com/senaite/senaite.core/blob/2.x/src/senaite/core/setuphandlers.py#L89

**Review by ramonski — APPROVED** [review]

**Pull request merged**